recipes/qt/all/conanfile.py

```diff
@@ -0,0 +1,673 @@
+        "with_icu": True,
+        "with_harfbuzz": False,
+        "with_libjpeg": True,
+        "with_libpng": True,
+        "with_sqlite3": True,
+        "with_mysql": True,
+        "with_pq": True,
+        "with_odbc": True,
+        "with_libalsa": False,
+        "with_openal": True,
+        "with_zstd": True,
+
+        "GUI": True,
+        "widgets": True,
+
+        "device": None,
+        "cross_compile": None,
+        "sysroot": None,
+        "config": None,
+        "multiconfiguration": False,
+    }
+    default_options.update({module: False for module in _submodules if module != 'qtbase'})
+
+    requires = "zlib/1.2.11"
+    short_paths = True
+
+    def build_requirements(self):
+        if tools.os_info.is_windows and self.settings.compiler == "Visual Studio":
+            self.build_requires("jom/1.1.3")
+        if self.options.qtwebengine:
+            self.build_requires("ninja/1.10.0")
+            # gperf, bison, flex, python >= 2.7.5 & < 3
+            if self.settings.os != "Windows":
+                if not tools.which("bison"):
+                    self.build_requires("bison/3.5.3")
+                if not tools.which("gperf"):
+                    self.build_requires("gperf/3.1")
+                if not tools.which("flex"):
+                    self.build_requires("flex/2.6.4")
+
+            # Check if a valid python2 is available in PATH or it will failflex
+            # Start by checking if python2 can be found
+            python_exe = tools.which("python2")
+            if not python_exe:
+                # Fall back on regular python
+                python_exe = tools.which("python")
+
+            if not python_exe:
+                msg = ("Python2 must be available in PATH "
+                       "in order to build Qt WebEngine")
+                raise ConanInvalidConfiguration(msg)
+            # In any case, check its actual version for compatibility
+            from six import StringIO  # Python 2 and 3 compatible
+            mybuf = StringIO()
+            cmd_v = "{} --version".format(python_exe)
+            self.run(cmd_v, output=mybuf)
+            verstr = mybuf.getvalue().strip().split('Python ')[1]
+            if verstr.endswith('+'):
+                verstr = verstr[:-1]
+            version = tools.Version(verstr)
+            # >= 2.7.5 & < 3
+            v_min = "2.7.5"
+            v_max = "3.0.0"
+            if (version >= v_min) and (version < v_max):
+                msg = ("Found valid Python 2 required for QtWebengine:"
+                       " version={}, path={}".format(mybuf.getvalue(), python_exe))
+                self.output.success(msg)
+            else:
+                msg = ("Found Python 2 in path, but with invalid version {}"
+                       " (QtWebEngine requires >= {} & < "
+                       "{})".format(verstr, v_min, v_max))
+                raise ConanInvalidConfiguration(msg)
+
+    def config_options(self):
```

Comment:
default options of Qt are platform dependent. For example `opengl=dynamic` should be the default on Windows (but not for others os, because it's not supported).

So I think that:
- in `config_options` method: a set of (working) default options based on os should be provided
- in `configure` method: raise any incompatible combinations of options, instead of silently overring options.

Reply:
the problem is then: how do you set default_option so that it does not raise by default on any platform ? The only solution I have in mind for that is to add an "auto" value for these options, make it the default, and set the value in configure depending on the settings(only when the option value is "auto" of course). I'm not sure if this kind of logic already exists in some CCI recipes. Would it be accepted ?

Reply:
AFAIK, recipe options set in `config_options` are default options, they don't override values set by downstream. So it's the good place for conditional default options.

So:
- in `default_options`: set default options which are the same for all settings
- in `config_options`: set default options constrained by settings (a combination which doesn't raise in `configure`).
- in `configure`: raise any invalid options combinations.

Reply:
Really ? I always thought that setting an option value in config_options was restraining the values available to the user. That is what is described in https://github.com/conan-io/conan/issues/3519#issuecomment-420211826

Reply:
Well, maybe I'm wrong about config_options behaviour, but a very recent recipe has implemented such logic, just can't remember which one 😞

Reply:
I was confused by this too a while ago https://github.com/conan-io/conan-center-index/issues/2574#issuecomment-674976187

Reading the comments in the old Conan issue confuses me now again 😄

Reply:
there are several recipes on CCI with option value `[True, False, "auto"]`. I guess it's the best solution for this currently

Reply:
From conan's documentation:

>  `config_options()` is used to configure or constraint the available options in a package, before they are given a value.

Setting a default value for an option is just part of configuring the available options of a package, so it fits within the confines of what is documented. If it is only for constraining the available options (e.g. by removing fPIC on windows), then conan's documentation is wrong and should be changed. However, in my experience, setting default values for options works exactly as expected, so I don't think there is an issue with conan's documentation.